src/js/player.js

```diff
@@ -2218,27 +2219,23 @@ class Player extends Component {
    *         The current video source when getting
    */
   src(source) {
-    if (source === undefined) {
-      return this.cache_.src;
-    }
-
-    this.changingSrc_ = true;
-
     let src = source;
 
-    if (Array.isArray(source)) {
-      this.cache_.sources = source;
-      src = source[0];
-    } else if (typeof source === 'string') {
-      src = {
-        src: source
-      };
+    // filter out invalid sources and turn our source into
+    // an array of source objects
+    src = filterSource(src);
 
-      this.cache_.sources = [src];
+    // no valid sources, return the current source
+    if (!src.length) {
+      return this.cache_.src;
     }
 
-    this.cache_.source = src;
+    this.cache_.sources = src;
+    src = src[0];
+
+    this.changingSrc_ = true;
 
+    this.cache_.source = src;
     this.currentType_ = src.type;
```

Comment:
If you have a source like this:
```js
{
  src: 'foo',
  type: 'video/foo'
}
```
With a middleware that returns an m3u8 for it, we want the following behavior:
```js
player.src();
// 'http://example.com/video.m3u8'

player.currentSrc();
// 'foo'

player.currentSource();
// {src: 'foo', type: 'video/foo'}
```

Let me double check what we talked about for `currentType` previously.